Clarify OnAckReceived pseudocode

Ralith opened the pull request:
I ran into a few points of confusion in the existing code:
- The time to update and scope of `largest_acked_packet` is not explicit
- `ProcessECN` references elements of `sent_packets` after they are removed by `OnPacketAcked`
- `DetectLostPackets` is passed an undefined variable

The fixes are trivial, except for the last: it isn't clear to me whether `DetectLostPackets` was intended to be called with `ack.largest_acked` or `largest_acked_packet`. These only differ when ACKs are reordered, and because the prior invocation of `DetectLostPackets` with a higher `largest_acked` would have already deemed lost all packets that the later invocation might deem lost, I don't think the behavior differs even then. Hence, removing the argument and consistently referencing `largest_acked_packet` seems to be the simpler option.

martinthomson commented on draft-ietf-quic-recovery.md:
```diff
@@ -720,9 +720,14 @@ Pseudocode for OnAckReceived and UpdateRtt follow:
     // ack-eliciting, update the RTT.
     if (sent_packets[ack.largest_acked] &&
         sent_packets[ack.largest_acked].ack_eliciting):
+      largest_acked_packet = ack.largest_acked
```
This doesn't work because the largest_acked in the ACK might be newly acknowledged, but not the largest ever acknowledged.

```suggestion
      if ack.largest_acked > largest_acked_packet:
         largest_acked_packet = ack.largest_acked
```

martinthomson commented on draft-ietf-quic-recovery.md:
```diff
@@ -720,9 +720,14 @@ Pseudocode for OnAckReceived and UpdateRtt follow:
 
+    // Process ECN information if present.
+    if (ACK frame contains ECN information):
+       ProcessECN(ack)
```
This is a separate change.  I think that you should make this a separate PR.

Ralith commented:
Split the `ProcessECN` tweak into https://github.com/quicwg/base-drafts/pull/2142, corrected the `largest_acked_packet` update logic, and added initialization for `largest_acked_packet`.

janaiyengar commented on draft-ietf-quic-recovery.md:
```diff
@@ -634,7 +634,7 @@ largest_sent_packet:
 
 largest_acked_packet:
-: The largest packet number acknowledged in an ACK frame.
+: The largest packet number acknowledged by any ACK frame.
```
```suggestion
: The largest packet number acknowledged in the connection so far.
```

martinthomson replied:
Isn't this in the packet number space?

martinthomson replied:
I think that you mean "packet number space" rather than connection.

ianswett commented on draft-ietf-quic-recovery.md:
```diff
@@ -716,6 +717,9 @@ Pseudocode for OnAckReceived and UpdateRtt follow:
 
 ~~~
   OnAckReceived(ack):
+    largest_acked_packet = max(largest_acked_packet,
```
I think this accurately describes the text we have, but I think we should add some text that you MAY use the largest acked from the most recent ACK, which could reduce spurious retransmits when one packet arrives far out of order?

ianswett replied:
To clarify, I'm ok doing this in a follow up PR.

Ralith replied:
> you MAY use the largest acked from the most recent ACK, which could reduce spurious retransmits when one packet arrives far out of order

Any reason not to make this the sole recommendation? I had some trouble coming up with a case where that might occur, which is why I didn't take that route originally, but I think I see it now: when a reordered ACK newly acknowledges a packet such that the newly acknowledged packet is > ack.largest_acked - kPacketThreshold, but <= largest_acked_packet - kPacketThreshold.

Ralith replied:
Er, wait, that's nonsense. Still having trouble seeing the case.

ianswett replied:
It's the case when one packet is far out of order, and then subsequent packets arrive in order.  ie: 100 is received, then 1, 2, 3, 4, 5, etc.  I don't fully understand the cause, but I've seen small UDP packets jump ahead of large ones in a number of traces.

But I suspect Jana and I will merge this as is and we can add follow-up text about that issue if we think it's valuable.
